File: src/main/java/cqrs/mr/ui/HomeController.java
package cqrs.mr.ui;

import cqrs.core.bus.CommandBus;
import cqrs.mr.commands.*;
import cqrs.mr.readModel.ReadModelFacade;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.ui.ModelMap;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.servlet.ModelAndView;

import java.util.UUID;

@Controller
public class HomeController {
	private final CommandBus bus;
	private final ReadModelFacade readmodel;

	@Autowired
	public HomeController(CommandBus bus, ReadModelFacade readModel) {
		this.bus = bus;
		this.readmodel = readModel;
	}

	@RequestMapping("/index.do")
	public String index(Model model) {
        model.addAttribute("inventoryItems", readmodel.getInventoryItems());
        return "index";
	}

	@RequestMapping("/details.do")
	public ModelAndView details(@RequestParam("id") String id) {
		UUID uuid = UUID.fromString(id);
        ModelMap map = new ModelMap();
        map.put("inventoryItem", readmodel.getInventoryItemDetails(uuid));
        return new ModelAndView("details", map);
	}



	@RequestMapping(value = "/add.do", method = RequestMethod.GET)
	public String add() {
        return "add";
	}

	@RequestMapping("/changeNamePage.do")
	public ModelAndView changeNamePage(@RequestParam("id") String id) {
		UUID uuid = UUID.fromString(id);
		ModelMap map = new ModelMap();
		map.put("inventoryItem", readmodel.getInventoryItemDetails(uuid));
		return new ModelAndView("rename", map);
	}

    @RequestMapping("/checkInPage.do")
    public ModelAndView checkInPage(@RequestParam("id") String id) {
        UUID uuid = UUID.fromString(id);
        ModelMap map = new ModelMap();
        map.put("inventoryItem", readmodel.getInventoryItemDetails(uuid));
        return new ModelAndView("checkin", map);
    }

	@RequestMapping("/removePage.do")
	public ModelAndView removePage(@RequestParam("id") String id) {
		UUID uuid = UUID.fromString(id);
        ModelMap map = new ModelMap();
        map.put("inventoryItem", readmodel.getInventoryItemDetails(uuid));
        return new ModelAndView("remove", map);
		
	}

	/*---- application state change ------------------------*/
    @RequestMapping(value = "/add.do", method = RequestMethod.POST)
    public ModelAndView add(@RequestParam("name") String name) {
        bus.send(new CreateInventoryItem(UUID.randomUUID(), name));
        return toIndexPage();
    }

    @RequestMapping(value = "/changeName.do", method = RequestMethod.POST)
    public ModelAndView changeName(@RequestParam("id") String id, @RequestParam("version") int version, @RequestParam("newName") String newName) {
        UUID uuid = UUID.fromString(id);
        bus.send(new RenameInventoryItem(uuid, newName, version));
        return toIndexPage();
    }

    @RequestMapping(value = "/deactivate.do", method = RequestMethod.POST)
    public ModelAndView deactivate(@RequestParam("id") String id, @RequestParam("version") int version) {
        UUID uuid = UUID.fromString(id);
        bus.send(new DeactivateInventoryItem(uuid, version));
        return toIndexPage();
    }

    @RequestMapping(value = "/checkIn.do", method = RequestMethod.POST)
    public ModelAndView checkIn(@RequestParam("id") String id, @RequestParam("version") int version, @RequestParam("count") int count) {
        UUID uuid = UUID.fromString(id);
        bus.send(new CheckInItemsToInventory(uuid, count, version));
        return toIndexPage();
    }

    @RequestMapping(value = "/remove.do", method = RequestMethod.POST)
    public ModelAndView remove(@RequestParam("id") String id, @RequestParam("version") int version, @RequestParam("count") int count) {
        UUID uuid = UUID.fromString(id);
        bus.send(new RemoveItemsFromInventory(uuid, count, version));
        return toIndexPage();
    }

    private ModelAndView toIndexPage() {
        ModelMap map = new ModelMap();
        map.put("inventoryItems", readmodel.getInventoryItems());
        return new ModelAndView("index", map);
    }
}

File: app/cqrs/mr/ui/HomeController.scala
package cqrs.mr.ui

import java.util
import java.util.UUID

import cqrs.core.bus.{CommandBus, EventBus}
import cqrs.core.eventstore.InMemoryEventStore
import cqrs.mr.commandhandlers._
import cqrs.mr.commands._
import cqrs.mr.domain.InventoryItem
import cqrs.mr.events._
import cqrs.mr.infra.RepositoryImpl
import cqrs.mr.readModel._
import javax.inject.Inject
import play.api.mvc.{AbstractController, ControllerComponents}

class HomeController @Inject()(cc: ControllerComponents) extends AbstractController(cc) {
  val eventBus = new EventBus()
  eventBus.registerHandler(classOf[InventoryItemCreated], DetailsView.createInventoryItemCreatedHandler())
  eventBus.registerHandler(classOf[InventoryItemDeactivated], DetailsView.createInventoryItemDeactivatedHandler())
  eventBus.registerHandler(classOf[InventoryItemRenamed], DetailsView.createInventoryItemRenamedHandler())
  eventBus.registerHandler(classOf[ItemsCheckedInToInventory], DetailsView.createItemsCheckedInToInventoryHandler())
  eventBus.registerHandler(classOf[ItemsRemovedFromInventory], DetailsView.createItemsRemovedFromInventoryHandler())

  //--event handlers: list view
  eventBus.registerHandler(classOf[InventoryItemCreated], ListView.createInventoryItemCreatedHandler())
  eventBus.registerHandler(classOf[InventoryItemRenamed], ListView.createInventoryItemRenamedHandler())
  eventBus.registerHandler(classOf[InventoryItemDeactivated], ListView.createInventoryItemDeactivatedHandler())

  val eventStore = new InMemoryEventStore(eventBus)
  val repo = new RepositoryImpl[InventoryItem](eventStore)

  val bus = new CommandBus()
  bus.registerHandler(classOf[CreateInventoryItem], new CreateInventoryItemHandler(repo))
  bus.registerHandler(classOf[CheckInItemsToInventory], new CheckInItemsToInventoryHandler(repo))
  bus.registerHandler(classOf[DeactivateInventoryItem], new DeactivateInventoryItemHandler(repo))
  bus.registerHandler(classOf[RemoveItemsFromInventory], new RemoveItemsFromInventoryHandler(repo))
  bus.registerHandler(classOf[RenameInventoryItem], new RenameInventoryItemHandler(repo))

  val readModel: ReadModelFacade = new ReadModelFacade()

  def index = Action {
    val items: util.List[InventoryItemListDto] = readModel.getInventoryItems()
    Ok(cqrs.mr.ui.html.index(items))
  }

  def showAddPage = Action {
    Ok(cqrs.mr.ui.html.add())
  }

  def detailsPage(id: String) = Action {
    val uuid = UUID.fromString(id)
    val item: InventoryItemDetailsDto = readModel.getInventoryItemDetails(uuid)
    Ok(cqrs.mr.ui.html.details(item))
  }

  def renamePage(id: String) = Action {
    val uuid = UUID.fromString(id)
    val item = readModel.getInventoryItemDetails(uuid)
    Ok(cqrs.mr.ui.html.rename(item))
  }

  def checkInPage(id: String) = Action {
    val uuid = UUID.fromString(id)
    val item = readModel.getInventoryItemDetails(uuid)
    Ok(cqrs.mr.ui.html.checkin(item))
  }

  def removePage(id: String) = Action {
    val uuid = UUID.fromString(id)
    val item = readModel.getInventoryItemDetails(uuid)
    Ok(cqrs.mr.ui.html.remove(item))
  }

  def add() = Action(parse.formUrlEncoded) { request =>
    val name = request.body("name").head
    bus.send(new CreateInventoryItem(UUID.randomUUID, name))
    Redirect("index.do")
  }

  def changeName() = Action(parse.formUrlEncoded) { request =>
    val id = request.body("id").head
    val name = request.body("newName").head
    val version: String = request.body("version").head
    bus.send(new RenameInventoryItem(UUID.fromString(id), name, version.toInt))
    Redirect("index.do")
  }

  def deactivate() = Action(parse.formUrlEncoded) { request =>
    val id = request.body("id").head
    val version = request.body("version").head
    bus.send(new DeactivateInventoryItem(UUID.fromString(id), version.toInt))
    Redirect("index.do")
  }

  def checkIn() = Action(parse.formUrlEncoded) { request =>
    val id = request.body("id").head
    val version = request.body("version").head
    val count = request.body("count").head
    bus.send(new CheckInItemsToInventory(UUID.fromString(id), count.toInt, version.toInt))
    Redirect("index.do")
  }

  def checkOut() = Action(parse.formUrlEncoded) { request =>
    val id = request.body("id").head
    val version = request.body("version").head
    val count = request.body("count").head

    bus.send(new RemoveItemsFromInventory(UUID.fromString(id), count.toInt, version.toInt))
    Redirect("index.do")
  }
}
